File: scripts/lib/mutation_inflight.py
#!/usr/bin/env python3
# SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
# SPDX-FileCopyrightText: Copyright (c) 2026 newmassrael
"""What a mutation round is doing RIGHT NOW, written where a later one can read it.

`scripts/lib/mutation_ledger.sh` records what a round CONCLUDED. This records
what it is in the middle of, and it exists for the half of a round's evidence a
verdict cannot carry: a round that is killed reaches no verdict, and — until
this file — reached no restore either.

`scripts/mutate` hangs every restore off one `trap ... EXIT`. A trap does not
run on SIGKILL, so a round killed outright leaves its last mutation in the
working tree. Both consequences read as success:

  - the next round takes that mutated file as its baseline, and whatever it
    then reports is a measurement of a tree nobody chose;
  - a successor session reads the edit as leftovers and reverts it by hand.
    Twice, measured. When the round was in fact still alive, that revert
    produced a false CAUGHT, which is a verdict that reads as a test working.

What made this unrecoverable was never lost data. `scripts/mutate` copies every
declared file before the first case, and that snapshot — the original bytes,
exactly — survives the kill intact under its `mktemp -d` directory. Only its
NAME dies, because the name lived nowhere but in the killed process's memory.

So the round writes the name down, before it touches a file, at a path with no
session in it: the same argument and the same root as the verdict ledger's.
The record is removed once the restore is done AND verified, so the records
still present are exactly the rounds that did not finish.

Four subcommands. `open` and `case` are called by a round, `close` from its
EXIT trap, and `recover` by the NEXT invocation of the harness — which is the
one that pays this file's debts, because it is the one that is still alive.

Output is tagged so the caller can speak in its own voice: each report line is
`<R|G|D><TAB><text>`, for the caller's red, green and dim. Exit status is 0
when nothing is outstanding and 1 when this file refuses to guess.
"""

import argparse
import hashlib
import json
import os
import shutil
import sys
import tempfile
import time

TAG_RED = "R"
TAG_GREEN = "G"
TAG_DIM = "D"


def emit(tag, text):
    sys.stdout.write("%s\t%s\n" % (tag, text))


def sha256(path):
    digest = hashlib.sha256()
    with open(path, "rb") as handle:
        for chunk in iter(lambda: handle.read(1 << 16), b""):
            digest.update(chunk)
    return digest.hexdigest()


def boot_id():
    """Which boot this is.

    A pid plus its start time is unique for as long as a boot lasts and not one
    moment longer, so the boot is part of the identity or the identity is a
    coincidence waiting to happen.
    """
    try:
        with open("/proc/sys/kernel/random/boot_id") as handle:
            return handle.read().strip()
    except OSError:
        return "unknown"


def proc_stat(pid):
    """The fields of /proc/<pid>/stat from the state onwards, or None.

    Sliced after the LAST `)` rather than counted from the front: the comm
    field is parenthesised and may itself contain spaces and parentheses,
    which is exactly what breaks a plain field count. What remains starts at
    field 3, so field N of the manual page is at index N - 3 here.
    """
    try:
        with open("/proc/%d/stat" % pid) as handle:
            raw = handle.read()
    except OSError:
        return None
    try:
        return raw.rsplit(") ", 1)[1].split()
    except IndexError:
        return None


def start_time(pid):
    """When a process started, in ticks since boot, or None if unreadable."""
    fields = proc_stat(pid)
    if fields is None or len(fields) < 20:
        return None
    return fields[19]


def process_state(pid):
    """The process's state letter, or None if unreadable."""
    fields = proc_stat(pid)
    if not fields:
        return None
    return fields[0]


def process_is_alive(record):
    """Is the process this record names still the process that wrote it?

    Every uncertain answer here is "alive", deliberately. Refusing to touch a
    round that has in fact died costs one message on the next invocation;
    reverting one that is in fact live is the failure this whole file exists to
    stop, and it produces a verdict rather than an error.
    """
    pid = record.get("pid")
    if not isinstance(pid, int) or pid <= 0:
        return True
    try:
        os.kill(pid, 0)
    except ProcessLookupError:
        return False
    except OSError:
        return True
    # A killed round whose parent has not reaped it is a zombie, and a zombie
    # answers `kill -0` and keeps its /proc entry. Reading that as "still
    # running" is the one way this check fails in the direction that leaves the
    # mutation in the tree forever, so the state letter is asked for by name.
    if process_state(pid) in ("Z", "X"):
        return False
    recorded = record.get("starttime")
    current = start_time(pid)
    if not recorded or not current:
        return True
    return record.get("boot_id") == boot_id() and recorded == current


def describe(record):
    """The round, in the words a reader needs to decide what it was."""
    what = record.get("casefile") or "(an unnamed casefile)"
    if record.get("mode") and record["mode"] != "run":
        what += " (--%s)" % record["mode"]
    if record.get("case"):
        what += ", applying %r" % record["case"]
    return what


def outstanding(record):
    """The declared paths whose bytes are not the ones the round started from.

    Compared against the hashes taken WITH the snapshot, not against git: a
    round is entitled to start from a dirty tree, and "back where it was" for
    such a round is that dirty content and not HEAD's.
    """
    left = []
    for target in record.get("targets") or []:
        path, want = target.get("path"), target.get("sha256")
        try:
            have = sha256(path)
        except OSError:
            have = None
        if have != want:
            left.append(path)
    return left


def discard_scratch(record):
    """Remove the round's `mktemp -d` directory — the `rm -rf` its trap missed.

    Only when the snapshot this file just read lives inside it, so a record
    naming some other directory removes nothing.
    """
    work = record.get("work") or ""
    snapshot = record.get("snapshot") or ""
    if not work or not snapshot:
        return
    if not snapshot.startswith(work.rstrip(os.sep) + os.sep):
        return
    shutil.rmtree(work, ignore_errors=True)


def write_record(path, record):
    """Replace the record in one step.

    A kill landing in the middle of a rewrite must leave the PREVIOUS record,
    not half of one: the record's whole worth is that it can be read by
    something that did not write it.
    """
    scratch = path + ".new"
    with open(scratch, "w") as handle:
        json.dump(record, handle, ensure_ascii=False)
        handle.write("\n")
        handle.flush()
        os.fsync(handle.fileno())
    os.replace(scratch, path)


def load(path):
    with open(path) as handle:
        return json.load(handle)


def retire(path):
    """Remove a record that has been accounted for.

    Tolerant of the file already being gone: two harnesses can start at once,
    and both are entitled to conclude that the same abandoned round has been
    put right.
    """
    try:
        os.unlink(path)
    except FileNotFoundError:
        pass


def cmd_open(args):
    """Write the record, and print the path it landed at.

    The NAME is chosen here rather than by a `mktemp` in the caller, because a
    caller that creates the file first leaves a zero-byte `.json` sitting in the
    directory until the content arrives — and a concurrent invocation that reads
    it in that instant sees not a record but a parse error, and refuses. Measured
    2026-08-26 running this repository's own harness suites five at a time.
    Staged under a name `recover` does not consider, and moved into place in one
    step, a record is either absent or complete.
    """
    targets = []
    for line in sys.stdin.read().splitlines():
        if not line:
            continue
        targets.append({"path": line, "sha256": sha256(line)})
    os.makedirs(args.dir, exist_ok=True)
    handle, staging = tempfile.mkstemp(prefix="round-", suffix=".json.new", dir=args.dir)
    os.close(handle)
    record = staging[: -len(".new")]
    write_record(
        record,
        {
            "schema": 1,
            "pid": args.pid,
            "starttime": start_time(args.pid),
            "boot_id": boot_id(),
            "host": os.uname().nodename,
            "repo": args.repo,
            "casefile": args.casefile,
            "mode": args.mode,
            "tree": args.tree,
            "work": args.work,
            "snapshot": args.snapshot,
            "started": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "targets": targets,
        },
    )
    # No cleanup: `write_record` stages at `<record>.new`, which IS the file
    # `mkstemp` just made, and moves it into place.
    print(record)
    return 0


def cmd_case(args):
    """Name the case now being applied, so a refusal can say which one it was.

    Which case a killed round died in is the difference between "something in
    this casefile is in your tree" and a sentence a reader can act on, and it is
    also the one case in the casefile that has no verdict.
    """
    try:
        record = load(args.record)
    except (OSError, ValueError):
        return 0
    record["case"] = args.label
    write_record(args.record, record)
    return 0


def cmd_close(args):
    """Remove the record — but only once the tree it named is back.

    The restore runs first, in the same EXIT trap, and `mutation_restore_quiet`
    swallows its errors by design. Removing the record unconditionally after it
    would hand the failure mode straight back: a restore that did not happen and
    no record left saying so. Keeping the record is also what tells the caller
    not to delete the snapshot the repair would need.
    """
    try:
        record = load(args.record)
    except OSError:
        return 0
    except ValueError as exc:
        emit(TAG_RED, "the in-flight record %s is unreadable (%s), so this round"
             % (args.record, exc))
        emit(TAG_RED, "cannot say whether it put the tree back. It is kept.")
        return 1
    left = outstanding(record)
    if not left:
        retire(args.record)
        return 0
    emit(TAG_RED, "the round ended with its mutation still in the tree:")
    for path in left:
        emit(TAG_RED, "  %s" % path)
    emit(TAG_RED, "the record that says how to put it back is kept at")
    emit(TAG_RED, "  %s" % args.record)
    emit(TAG_RED, "and so is the snapshot it names. The next `scripts/mutate` in this")
    emit(TAG_RED, "tree restores from it.")
    return 1


def cmd_recover(args):
    """Find the rounds that did not finish, and finish their restore.

    Runs before this invocation touches anything, because a leftover mutation
    that reaches the baseline is not a stale file — it is a measurement of the
    wrong tree, reported with the same confidence as any other.
    """
    if not os.path.isdir(args.dir):
        return 0
    refused = False
    for name in sorted(os.listdir(args.dir)):
        if not name.endswith(".json"):
            continue
        path = os.path.join(args.dir, name)
        try:
            record = load(path)
        except FileNotFoundError:
            # Another invocation repaired this one between the listing and the
            # read. A record that is GONE is the outcome this asks for, not a
            # record that cannot be read, and refusing here would turn two
            # harnesses running at once into a failure of both.
            continue
        except (OSError, ValueError) as exc:
            emit(TAG_RED, "an in-flight round record cannot be read: %s" % path)
            emit(TAG_RED, "  %s" % exc)
            emit(TAG_RED, "It is left where it is: an unreadable record is still the"
                 " only trace of")
            emit(TAG_RED, "whatever round wrote it, and deleting it here would leave"
                 " none.")
            refused = True
            continue
        # Another checkout's round. Its own harness owns that restore, and this
        # one has neither its snapshot's tree nor any business in it.
        if record.get("repo") != args.repo:
            continue
        # A round that is still running is not this invocation's business, and
        # saying so here would be wrong twice over: it is not a repair, and it
        # would make one harness's output depend on what else happens to be
        # running — which is how five concurrent `--check` runs came to fail on
        # each other's notes. Whoever wants to know asks
        # `scripts/mutation-ledger in-flight`, which exists for that question.
        if process_is_alive(record):
            continue
        if not handle_abandoned(path, record):
            refused = True
    return 1 if refused else 0


def cmd_list(args):
    """Every record, and whether its round is still going.

    The question `scripts/mutate` deliberately does not answer. A round in
    flight is not a repair, so reporting it from a harness that is trying to
    start one makes that harness's output depend on what else is running — but
    the question itself is the one a session opens with, because a dirty tree
    that belongs to a LIVE round has been reverted by a successor twice.
    """
    if not os.path.isdir(args.dir):
        return 0
    for name in sorted(os.listdir(args.dir)):
        if not name.endswith(".json"):
            continue
        path = os.path.join(args.dir, name)
        try:
            record = load(path)
        except (OSError, ValueError):
            print("unreadable\t%s" % path)
            continue
        print(
            "%s\tpid=%s\t%s\t%s\t%s"
            % (
                "running" if process_is_alive(record) else "abandoned",
                record.get("pid"),
                record.get("started"),
                describe(record),
                record.get("repo"),
            )
        )
    return 0


def handle_abandoned(path, record):
    """One record whose round is gone. True when nothing is left outstanding."""
    left = outstanding(record)
    if not left:
        retire(path)
        discard_scratch(record)
        emit(TAG_DIM, "cleared the record of a round that died without leaving a"
             " mutation behind: %s" % describe(record))
        return True

    snapshot = record.get("snapshot") or ""
    restorable, unreachable = [], []
    for target in record.get("targets") or []:
        if target.get("path") not in left:
            continue
        source = os.path.join(snapshot, target["path"])
        try:
            if sha256(source) == target.get("sha256"):
                restorable.append((target["path"], source, target["sha256"]))
                continue
        except OSError:
            pass
        unreachable.append(target["path"])

    if unreachable:
        emit(TAG_RED, "a mutation round was killed in this tree and its mutation is"
             " still here:")
        for still_there in left:
            emit(TAG_RED, "  %s" % still_there)
        emit(TAG_RED, "the round was %s" % describe(record))
        emit(TAG_RED, "and the bytes it started from are not where it said they would"
             " be:")
        emit(TAG_RED, "  %s" % (snapshot or "(the record names no snapshot)"))
        emit(TAG_RED, "This harness will not guess at them. `git checkout` would put"
             " back HEAD,")
        emit(TAG_RED, "which is not what the round started from whenever the tree was"
             " dirty, and")
        emit(TAG_RED, "would discard whatever else those files carried. Put them back"
             " by hand and")
        emit(TAG_RED, "run again: this refusal clears itself once their bytes match"
             " what the round")
        emit(TAG_RED, "recorded. The record is kept at %s" % path)
        return False

    for target_path, source, want in restorable:
        with open(source, "rb") as handle:
            original = handle.read()
        # Into the existing file, so its mode survives — the reason
        # `mutation_restore` copies rather than moves.
        with open(target_path, "wb") as handle:
            handle.write(original)
        if sha256(target_path) != want:
            emit(TAG_RED, "restoring %s did not reproduce the bytes the round"
                 " recorded." % target_path)
            emit(TAG_RED, "The record is kept at %s" % path)
            return False

    emit(TAG_GREEN, "restored %d file(s) left mutated by a round that was killed,"
         " from that round's own snapshot:" % len(restorable))
    for target_path, _, _ in restorable:
        emit(TAG_GREEN, "  %s" % target_path)
    emit(TAG_DIM, "  the round was %s" % describe(record))
    emit(TAG_DIM, "  started %s, pid %s, and reached no verdict"
         % (record.get("started"), record.get("pid")))
    retire(path)
    discard_scratch(record)
    return True


def main(argv):
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    sub = parser.add_subparsers(dest="command", required=True)

    opener = sub.add_parser("open", help="write the record; targets on stdin")
    opener.add_argument("--dir", required=True)
    opener.add_argument("--repo", required=True)
    opener.add_argument("--casefile", required=True)
    opener.add_argument("--mode", required=True)
    opener.add_argument("--tree", required=True)
    opener.add_argument("--work", required=True)
    opener.add_argument("--snapshot", required=True)
    opener.add_argument("--pid", required=True, type=int)
    opener.set_defaults(func=cmd_open)

    case = sub.add_parser("case", help="name the case now being applied")
    case.add_argument("--record", required=True)
    case.add_argument("--label", required=True)
    case.set_defaults(func=cmd_case)

    close = sub.add_parser("close", help="remove the record once the tree is back")
    close.add_argument("--record", required=True)
    close.set_defaults(func=cmd_close)

    recover = sub.add_parser("recover", help="finish the restore of rounds that died")
    recover.add_argument("--dir", required=True)
    recover.add_argument("--repo", required=True)
    recover.set_defaults(func=cmd_recover)

    listing = sub.add_parser("list", help="one line per record, running or not")
    listing.add_argument("--dir", required=True)
    listing.set_defaults(func=cmd_list)

    args = parser.parse_args(argv)
    return args.func(args)


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: sce-build/tests/mutation_inflight.rs
// SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2026 newmassrael
//
// A mutation round that is killed outright does not leave its mutation behind.
//
// `scripts/mutate` hangs every restore off one `trap ... EXIT`, and a trap does
// not run on SIGKILL. Measured twice: a round died with the tree still holding
// its mutation, the next round read that file as its baseline, and a successor
// session read it as leftovers and reverted it by hand — producing a false
// CAUGHT the second time, which is a verdict that reads as a test working.
//
// The repair is a record: `scripts/lib/mutation_inflight.py` writes what the
// round is about to mutate, and where the original bytes live, to a path with
// no session in it — then the NEXT invocation finishes the restore. Every
// failure of that record reads as success, which is why each one is a test:
//
//   - a record path that moves with whoever runs the round is a session id by
//     another name, and the next invocation finds nothing to repair;
//   - a LIVE round's edits look exactly like an abandoned round's. Reverting
//     them is the failure this whole record exists to stop, so "is it still
//     running" is asked of the kernel and every uncertain answer is "alive";
//   - a snapshot that did not survive must be NAMED, not guessed at. A
//     `git checkout` here would put back HEAD, which is not what a round
//     starting from a dirty tree began with;
//   - and a round that ended without putting the tree back must keep its
//     record, or the harness deletes the evidence of its own failure.
//
// The library is driven the way `scripts/mutate` drives it, so what is measured
// is the code that runs in a round and not a copy of it.

use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use tempfile::{tempdir, TempDir};

fn repo_root() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("sce-build has a parent directory")
        .to_path_buf()
}

fn tool() -> String {
    repo_root()
        .join("scripts/lib/mutation_inflight.py")
        .display()
        .to_string()
}

fn library() -> String {
    repo_root()
        .join("scripts/lib/mutation_ledger.sh")
        .display()
        .to_string()
}

/// Run a snippet from inside `cwd`, and return its stdout and whether it
/// succeeded. Both halves matter here: a refusal is a non-zero status AND the
/// sentence that says what it refused to guess at.
fn sh(cwd: &Path, script: &str) -> (String, bool) {
    let out = Command::new("bash")
        .arg("-c")
        .arg(script)
        .current_dir(cwd)
        .output()
        .expect("run bash");
    assert!(
        String::from_utf8_lossy(&out.stderr).trim().is_empty()
            || !String::from_utf8_lossy(&out.stderr).contains("Traceback"),
        "the tool died rather than answering:\n{}",
        String::from_utf8_lossy(&out.stderr)
    );
    (
        String::from_utf8_lossy(&out.stdout).into_owned(),
        out.status.success(),
    )
}

/// A tree standing in for a checkout mid-round: one declared file, a snapshot
/// holding the bytes it started from, and somewhere for the record to live.
struct Round {
    home: TempDir,
    repo: PathBuf,
    work: PathBuf,
    records: PathBuf,
}

impl Round {
    fn new() -> Self {
        let home = tempdir().expect("temp home");
        let repo = home.path().join("repo");
        let work = home.path().join("work");
        let records = home.path().join("records");
        fs::create_dir_all(work.join("snapshot")).expect("snapshot dir");
        fs::create_dir_all(&records).expect("records dir");
        fs::create_dir_all(&repo).expect("repo dir");
        Command::new("git")
            .args(["init", "-q", "."])
            .current_dir(&repo)
            .status()
            .expect("git init");
        fs::write(repo.join("f.txt"), "original\n").expect("write target");
        fs::write(work.join("snapshot/f.txt"), "original\n").expect("write snapshot");
        Round {
            home,
            repo,
            work,
            records,
        }
    }

    /// Write the record the way a round does, and return where it landed. The
    /// tool mints the name, so a caller cannot leave a half-written one behind.
    ///
    /// `alive` decides whose pid the record carries: this test process, which
    /// is running, or a shell that exits the moment it has written — an
    /// abandoned round, exactly.
    fn open(&self, alive: bool) -> PathBuf {
        let pid = if alive {
            format!("{}", std::process::id())
        } else {
            "$$".to_string()
        };
        let script = format!(
            "printf 'f.txt\\n' | bash -c 'exec python3 {tool} open \
             --dir {dir} --repo {repo} --casefile demo.cases --mode run \
             --tree deadbeef --work {work} --snapshot {work}/snapshot --pid {pid}'",
            tool = tool(),
            dir = self.records.display(),
            repo = self.repo.display(),
            work = self.work.display(),
        );
        let (out, ok) = sh(&self.repo, &script);
        assert!(ok, "opening the record failed: {out}");
        let record = PathBuf::from(out.trim());
        assert!(
            record.is_file(),
            "the tool did not print where the record landed: {out}"
        );
        let (out, ok) = sh(
            &self.repo,
            &format!(
                "python3 {} case --record {} --label 'a demo case'",
                tool(),
                record.display()
            ),
        );
        assert!(ok, "naming the case failed: {out}");
        record
    }

    fn mutate(&self) {
        fs::write(self.repo.join("f.txt"), "MUTATED\n").expect("apply the mutation");
    }

    fn target(&self) -> String {
        fs::read_to_string(self.repo.join("f.txt")).expect("read the target")
    }

    fn recover(&self) -> (String, bool) {
        sh(
            &self.repo,
            &format!(
                "python3 {} recover --dir {} --repo {}",
                tool(),
                self.records.display(),
                self.repo.display()
            ),
        )
    }

    /// Only the records themselves: the tool stages content beside them under
    /// a name it does not consider, and counting that would be counting a
    /// halfway state as a round.
    fn records_left(&self) -> usize {
        fs::read_dir(&self.records)
            .expect("read records dir")
            .filter_map(|entry| entry.ok())
            .filter(|entry| entry.file_name().to_string_lossy().ends_with(".json"))
            .count()
    }
}

#[test]
fn the_in_flight_record_does_not_move_with_the_harness_that_runs_the_round() {
    let home = tempdir().expect("temp home");
    // `XDG_DATA_HOME` points somewhere that is not the answer, because the loop
    // harness driving these rounds exports its own — and the verdict ledger's
    // first shape consulted it, which put the very first round's record under a
    // directory named after the run rather than after the corpus.
    let out = Command::new("bash")
        .arg("-c")
        .arg(format!(
            "set -euo pipefail; source {}; mutation_inflight_dir",
            library()
        ))
        .current_dir(repo_root())
        .env("HOME", home.path())
        .env("XDG_DATA_HOME", home.path().join("decoy"))
        .env_remove("SCE_MUTATION_INFLIGHT_DIR")
        .output()
        .expect("ask the library where the record goes");
    let dir = String::from_utf8_lossy(&out.stdout).trim().to_string();

    assert_eq!(
        dir,
        home.path()
            .join(".local/share/sce-mutation-corpus/in-flight")
            .display()
            .to_string(),
        "the record's path is derived from something other than HOME, so it \
         moves with whoever runs the round — and the next invocation, which is \
         the one that can repair the tree, looks somewhere else"
    );
    assert!(
        !dir.contains("decoy"),
        "the record followed XDG_DATA_HOME: {dir}"
    );
}

#[test]
fn a_killed_rounds_mutation_is_put_back_from_its_own_snapshot() {
    let round = Round::new();
    round.open(false);
    round.mutate();
    assert_eq!(round.target(), "MUTATED\n", "the setup did not mutate");

    let (report, ok) = round.recover();

    assert!(ok, "the recovery refused a repair it could make:\n{report}");
    assert_eq!(
        round.target(),
        "original\n",
        "the mutation a killed round left behind is still in the tree:\n{report}"
    );
    assert!(
        report.contains("f.txt"),
        "the repair did not say what it put back:\n{report}"
    );
    // Which round, and which case. A killed round's case is the one case that
    // reached no verdict, and a report that cannot name it sends the reader
    // looking through the whole casefile.
    assert!(
        report.contains("demo.cases") && report.contains("a demo case"),
        "the repair did not say whose mutation it was:\n{report}"
    );
    assert_eq!(
        round.records_left(),
        0,
        "the record outlived the repair, so the next invocation repeats it"
    );
    assert!(
        !round.work.exists(),
        "the round's scratch survived, which is the `rm -rf` its trap missed"
    );
}

#[test]
fn a_round_that_is_still_running_is_left_alone() {
    let round = Round::new();
    // The record carries this test process's pid, so the round it names is
    // running by construction.
    round.open(true);
    round.mutate();

    let (report, ok) = round.recover();

    assert!(ok, "a live round was treated as a failure:\n{report}");
    assert_eq!(
        round.target(),
        "MUTATED\n",
        "a LIVE round's mutation was reverted. That is the measured failure \
         this record exists to stop: reverted before the build it reads as \
         SURVIVED, reverted during it as CAUGHT, and the second reads as a \
         test doing its job:\n{report}"
    );
    assert_eq!(
        round.records_left(),
        1,
        "the live round's record was removed, so nothing will repair its tree \
         if it is killed next"
    );
    // And silent, deliberately. A round in flight is not a repair, and saying
    // so from a harness that is trying to START a round makes that harness's
    // output depend on what else is running — measured 2026-08-26, five
    // concurrent `--check` runs failing on each other's notes.
    assert!(
        report.trim().is_empty(),
        "a harness starting its own round reported on someone else's:\n{report}"
    );

    // The question itself is not dropped, it is moved: this is what a session
    // asks before it decides a dirty tree is leftovers.
    let (listing, ok) = sh(
        &round.repo,
        &format!("python3 {} list --dir {}", tool(), round.records.display()),
    );
    assert!(ok, "the listing failed:\n{listing}");
    assert!(
        listing.contains("running") && listing.contains("a demo case"),
        "nothing can tell a session that this tree's edits belong to a live \
         round, which is how two of them came to be reverted:\n{listing}"
    );
}

#[test]
fn a_snapshot_that_did_not_survive_is_named_rather_than_guessed_at() {
    let round = Round::new();
    round.open(false);
    round.mutate();
    fs::remove_dir_all(&round.work).expect("lose the snapshot");

    let (report, ok) = round.recover();

    assert!(
        !ok,
        "a repair it could not make was reported as done:\n{report}"
    );
    assert_eq!(
        round.target(),
        "MUTATED\n",
        "the file was changed without the bytes to change it back — a guess, \
         which is what `git checkout` would have been:\n{report}"
    );
    assert!(
        report.contains("f.txt") && report.contains("demo.cases"),
        "the refusal did not name what is still mutated:\n{report}"
    );
    assert_eq!(
        round.records_left(),
        1,
        "the record was removed along with the refusal, leaving no trace of a \
         tree that is still mutated"
    );

    // And it clears itself once the bytes are back, so the refusal is a state
    // the tree can leave rather than a flag someone has to remember to reset.
    fs::write(round.repo.join("f.txt"), "original\n").expect("put it back by hand");
    let (report, ok) = round.recover();
    assert!(
        ok,
        "the refusal outlived the condition that caused it:\n{report}"
    );
    assert_eq!(round.records_left(), 0, "the record survived its repair");
}

#[test]
fn a_record_from_another_checkout_is_not_acted_on() {
    let round = Round::new();
    // Same relative path, a different tree. `scripts/mutate --check` runs over
    // this corpus on every push, and a record naming another checkout must not
    // make one tree restore a file on behalf of another.
    fs::write(
        round.records.join("foreign.json"),
        format!(
            "{{\"schema\":1,\"pid\":1,\"starttime\":\"1\",\"boot_id\":\"x\",\
              \"repo\":\"/somewhere/else\",\"casefile\":\"other.cases\",\"mode\":\"run\",\
              \"work\":\"/nope\",\"snapshot\":\"/nope/snapshot\",\
              \"targets\":[{{\"path\":\"f.txt\",\"sha256\":\"{}\"}}]}}",
            "0".repeat(64)
        ),
    )
    .expect("write a foreign record");
    round.mutate();

    let (report, ok) = round.recover();

    assert!(ok, "another checkout's record failed this tree:\n{report}");
    assert_eq!(
        round.target(),
        "MUTATED\n",
        "a record belonging to another checkout was acted on here:\n{report}"
    );
    assert_eq!(
        round.records_left(),
        1,
        "another checkout's record was deleted by this tree's harness"
    );
}

#[test]
fn a_round_that_did_not_put_the_tree_back_keeps_its_record() {
    let round = Round::new();

    // Through the shell wrappers this time, because it is the EXIT trap's
    // contract that is under test: `mutation_restore_quiet` swallows its own
    // errors, so a restore that did not happen must be caught by the close —
    // and its non-zero status is what stops the trap deleting the snapshot the
    // repair will need.
    let open = format!(
        "set -euo pipefail; source {lib}; \
         mutation_inflight_open demo.cases run {work} {work}/snapshot 'f.txt'; \
         printf 'MUTATED\\n' > f.txt; \
         mutation_inflight_close && echo CLOSED_CLEAN || echo CLOSED_REFUSED",
        lib = library(),
        work = round.work.display(),
    );
    let (report, ok) = Command::new("bash")
        .arg("-c")
        .arg(&open)
        .current_dir(&round.repo)
        .env("HOME", round.home.path())
        .env("SCE_MUTATION_INFLIGHT_DIR", &round.records)
        .output()
        .map(|out| {
            (
                String::from_utf8_lossy(&out.stdout).into_owned(),
                out.status.success(),
            )
        })
        .expect("drive the shell wrappers");

    assert!(ok, "the wrappers failed outright:\n{report}");
    assert!(
        report.contains("CLOSED_REFUSED"),
        "a round whose tree is still mutated reported a clean close, which is \
         what lets the trap delete the snapshot:\n{report}"
    );
    assert!(
        report.contains("f.txt"),
        "the close did not name what was left in the tree:\n{report}"
    );
    assert_eq!(
        round.records_left(),
        1,
        "the record was removed by a close that had nothing to close"
    );

    // And the same wrappers on a tree that IS back leave nothing behind.
    fs::write(round.repo.join("f.txt"), "original\n").expect("put it back");
    let clean = format!(
        "set -euo pipefail; source {lib}; \
         mutation_inflight_open demo.cases run {work} {work}/snapshot 'f.txt'; \
         mutation_inflight_close && echo CLOSED_CLEAN",
        lib = library(),
        work = round.work.display(),
    );
    let out = Command::new("bash")
        .arg("-c")
        .arg(&clean)
        .current_dir(&round.repo)
        .env("HOME", round.home.path())
        .env("SCE_MUTATION_INFLIGHT_DIR", &round.records)
        .output()
        .expect("drive the shell wrappers");
    assert!(
        String::from_utf8_lossy(&out.stdout).contains("CLOSED_CLEAN"),
        "a round that put the tree back could not close its record:\n{}{}",
        String::from_utf8_lossy(&out.stdout),
        String::from_utf8_lossy(&out.stderr)
    );
    assert_eq!(
        round.records_left(),
        1,
        "the clean close left its own record behind (the one still here is the \
         refused one from the first half)"
    );
}
